### PR TITLE
UAR-618: Display individual BOs DOB

### PR DESCRIPTION
**Pull request by bmcgreevyCH**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-618

### Change description
Display individual BOs Date Of Birth

### CODE REVIEW

File: src/utils/update/psc.to.beneficial.owner.type.mapper.ts [hunk 1]
```diff
@@ -158,12 +158,19 @@ const natureOfControlTypeMap = new Map<string, NatureOfControlType>([
...
-  for (const private_bo_data of boPrivateData) {
-    if (private_bo_data.hashedId === ch_reference) {
-      return hasOffice
-        ? mapBOMOAddress(private_bo_data.principalAddress)
-        : mapBOMOAddress(private_bo_data.usualResidentialAddress);
+export const mapIndividualBOPrivateData = (boPrivateData: BeneficialOwnerPrivateData[], beneficialOwner: BeneficialOwnerIndividual) => {
+  for (const boData of boPrivateData) {
+    if (boData.hashedId === beneficialOwner.ch_reference) {
+      beneficialOwner.usual_residential_address = mapBOMOAddress(boData.usualResidentialAddress);
+      beneficialOwner.date_of_birth = mapInputDate(boData.dateOfBirth);
```
Comment (n-r-b) [inline src/utils/update/psc.to.beneficial.owner.type.mapper.ts:165]: Are there any views to update? We may have been setting the DOB day to empty on the check your answers screen

**Review by n-r-b — COMMENTED** [review]
Can you update the title of the PR to include the ticket number please?

File: src/utils/update/psc.to.beneficial.owner.type.mapper.ts (hunk 1 above)
Comment (bmcgreevyCH) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:165]: > Are there any views to update? We may have been setting the DOB day to empty on the check your answers screen

views/update/review-beneficial-owner-individual.html uses the same file I changed for MOs: views/includes/date_of_birth-display.html

Comment (bmcgreevyCH) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:165]: > I think AC 3 mentions that in **general**, if no DOB is retrieved then we display month and year @n-r-b @bmcgreevyCH

I spoke to Andre about this and D.O.B should be in CHIPS db. We have no way to check that the day is "missing".

File: src/utils/update/fetch.beneficial.owners.private.data.ts [hunk 2]
```diff
@@ -31,20 +31,21 @@ export const fetchBeneficialOwnersPrivateData = async (appData: ApplicationData,
 };
 
 const mapBeneficialOwnersPrivateData = (boPrivateData, appData: ApplicationData) => {
-  if (boPrivateData === undefined || boPrivateData.length === 0) {
-    return;
+  if (boPrivateData !== undefined && boPrivateData.length > 0) {
```
[the remark this answers is not in the document]
Comment (bmcgreevyCH) [reply src/utils/update/fetch.beneficial.owners.private.data.ts:34]: removing

File: src/utils/update/fetch.beneficial.owners.private.data.ts [hunk 3]
```diff
@@ -31,20 +31,19 @@ export const fetchBeneficialOwnersPrivateData = async (appData: ApplicationData,
...
     }
-  };
-  const mapPrincipalAddress = beneficialOwner => {
+  });
+  appData.update?.review_beneficial_owners_government_or_public_authority?.forEach(beneficialOwner => {
     if (beneficialOwner.ch_reference) {
-      beneficialOwner.principal_address = mapBoPrivateAddress(boPrivateData, beneficialOwner.ch_reference, true);
+      mapCorporateOrGovernmentBOPrivateData(boPrivateData, beneficialOwner);
     }
-  };
```
Comment (AndreKramer7654) [inline src/utils/update/fetch.beneficial.owners.private.data.ts:46]: we did these with closures due to duplicate code smell

File: src/utils/update/fetch.beneficial.owners.private.data.ts [hunk 4]
```diff
@@ -31,20 +31,20 @@ export const fetchBeneficialOwnersPrivateData = async (appData: ApplicationData,
...
-  appData.update?.review_beneficial_owners_individual?.forEach(mapUsualResidentialAddress);
-  appData.update?.review_beneficial_owners_government_or_public_authority?.forEach(mapPrincipalAddress);
-  appData.update?.review_beneficial_owners_corporate?.forEach(mapPrincipalAddress);
+  });
+  appData.update?.review_beneficial_owners_government_or_public_authority?.forEach(beneficialOwner => {
```
Comment (olaych) [inline src/utils/update/fetch.beneficial.owners.private.data.ts:44]: Apologies if I previously asked. Is there a test for `review_beneficial_owners_government_or_public_authority` been undefined? 
Same for the other BO review

Comment (bmcgreevyCH) [reply src/utils/update/fetch.beneficial.owners.private.data.ts:44]: Added some tests.

File: src/utils/update/psc.to.beneficial.owner.type.mapper.ts [hunk 5]
```diff
@@ -158,12 +158,19 @@ const natureOfControlTypeMap = new Map<string, NatureOfControlType>([
...
 ]);
 
-export const mapBoPrivateAddress = (boPrivateData: BeneficialOwnerPrivateData[], ch_reference: string, hasOffice: boolean) => {
-  for (const private_bo_data of boPrivateData) {
-    if (private_bo_data.hashedId === ch_reference) {
-      return hasOffice
-        ? mapBOMOAddress(private_bo_data.principalAddress)
-        : mapBOMOAddress(private_bo_data.usualResidentialAddress);
+export const mapIndividualBOPrivateData = (boPrivateData: BeneficialOwnerPrivateData[], beneficialOwner: BeneficialOwnerIndividual) => {
+  for (const boData of boPrivateData) {
```
[the remark this answers is not in the document]
Comment (bmcgreevyCH) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:162]: fetchBeneficialOwnersPrivateData does it before it gets here:
if (!boPrivateData || boPrivateData.length === 0) {
  logger.info(`No private Beneficial Owner details were retrieved for overseas entity ${appData.entity_number}`);
}

File: src/utils/update/fetch.beneficial.owners.private.data.ts (hunk 3 above)
Comment (bmcgreevyCH) [reply src/utils/update/fetch.beneficial.owners.private.data.ts:46]: Yeah I changed it back....Draconian Sonarqube is what that is. For all it was.

Comment (AndreKramer7654) [reply src/utils/update/fetch.beneficial.owners.private.data.ts:46]: 👍

File: src/utils/update/psc.to.beneficial.owner.type.mapper.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndreKramer7654) [reply src/utils/update/psc.to.beneficial.owner.type.mapper.ts:162]: And I see the type info now too.

**Review by AndreKramer7654 — APPROVED** [review]

**Review by azafarCH — APPROVED** [review]

File: src/utils/update/fetch.beneficial.owners.private.data.ts [hunk 6]
```diff
@@ -31,20 +31,18 @@ export const fetchBeneficialOwnersPrivateData = async (appData: ApplicationData,
...
 
 const mapBeneficialOwnersPrivateData = (boPrivateData, appData: ApplicationData) => {
-  if (boPrivateData === undefined || boPrivateData.length === 0) {
-    return;
-  }
-  const mapUsualResidentialAddress = beneficialOwner => {
+  const mapIndividualOwners = beneficialOwner => {
     if (beneficialOwner.ch_reference) {
-      beneficialOwner.usual_residential_address = mapBoPrivateAddress(boPrivateData, beneficialOwner.ch_reference, false);
+      mapIndividualBOPrivateData(boPrivateData, beneficialOwner);
```
Comment (olaych) [inline src/utils/update/fetch.beneficial.owners.private.data.ts:36]: I believe this value replaces the mapped value  of 
https://github.com/companieshouse/overseas-entities-web/blob/47a7c715f598289808ab49f8971c142af80434d9/src/utils/update/psc.to.beneficial.owner.type.mapper.ts#L20C5-L20C5

**Pull request merged**